### ISSUE TITLE
Spring Security 5.7.8 -> 6.0 method name renaming

### ISSUE THREAD

**Issue by JeremyTwiggs** [openrewrite/rewrite-spring#350]
Hi,

I've been trying to upgrade a Spring Boot 2.7.11 project to 3.0.x. and have noticed that the following methods do not get renamed:
```java
CsrfConfigurer#ignoringAntMatchers -> ignoringRequestMatchers
AuthorizationManagerRequestMatcherRegistry#mvcMatchers -> requestMatchers
AbstractRequestMatcherRegistry#antMatchers -> requestMatchers
```

These methods come from classes in spring-security. 

If it helps any, I searched your code base and found something that maybe tries to upgrde these, but (and I may be wrong) I'd have to guess that it works when upgrading spring-security from 5.8 to 6.0, however, by default, spring-boot 2.7.11 still uses spring-security 5.7.8 so maybe its not activated 🤔 

Any help would be appreciated.

**Comment by timtebeek** [comment]
Hi @JeremyTwiggs ; This was indeed a concern as we worked through the implementation of that recipe, with some of [the discussion captured as comments on the PR](https://github.com/openrewrite/rewrite-spring/pull/286), and other parts discussed over Slack.

From what I understand on your end it's not yet working as you'd hoped?

I'm thinking what we could do in this case; perhaps we need to introduce an explicit recipe to migrate to Spring Security 5.8 for those on Spring Boot 2.7.x, which includes this migration step of `requestMatchers`, and only then migrate people over to 3.x.

**Comment by joanvr** [comment]
@timtebeek do you think adding `UpgradeSpringSecurity_5_8` to `UpgradeSpringBoot_2_7` (or maybe 3?) could be a good solution for this issue? I've seen that `UpgradeSpringSecurity_5_8` has `UseNewRequestMatchers` in it, so aside from some unwanted side-effects, seems like a good match.

**Comment by timtebeek** [comment]
Yes I'd considered something along those lines as well indeed. Some thoughts below.

I'd maybe not add it to `UpgradeSpringBoot_2_7`, as it's an _optional_ step for folks on 2.7.x; we wouldn't want to push anything to them that's not required there, and could potentially break their application. As it stands it's currently a recipe that folks can run separately, which was fine up to now.

## What we could do
When going to 3.x however it suddenly becomes a requirement to pick up the changes introduced in `UpgradeSpringSecurity_5_8`. So perhaps it's then best to change `UpgradeSpringBoot_3_0` to include `UpgradeSpringSecurity_5_8`, such that these lines:
https://github.com/openrewrite/rewrite-spring/blob/3b939a638fcc815dbbda515e9c238f09fb0e5bd0/src/main/resources/META-INF/rewrite/spring-boot-30.yml#L28-L30
become
```diff
 recipeList:
   - org.openrewrite.java.spring.boot2.UpgradeSpringBoot_2_7
+  - org.openrewrite.java.spring.security5.UpgradeSpringSecurity_5_8
   - org.openrewrite.java.spring.boot3.RemoveEnableBatchProcessing
```
Note how I'm intentionally puttin that Spring Security 5.8 upgrade pretty early while the application is technically still a Spring Boot 2.7.x app, with corresponding dependencies.

## What we already do
Now we already include `UpgradeSpringSecurity_6_0` in `UpgradeSpringBoot_3_0` further down
https://github.com/openrewrite/rewrite-spring/blob/3b939a638fcc815dbbda515e9c238f09fb0e5bd0/src/main/resources/META-INF/rewrite/spring-boot-30.yml#L65-L66

And `UpgradeSpringSecurity_6_0` already calls out to `UpgradeSpringSecurity_5_8`.
https://github.com/openrewrite/rewrite-spring/blob/3b939a638fcc815dbbda515e9c238f09fb0e5bd0/src/main/resources/META-INF/rewrite/spring-security-60.yml#L27-L28

But perhaps these changes are either not applied because of an ordering issue with the recipes and dependencies being changed before we get to the Spring Security recipes, leading to mismatches. We might just want to verify the use case of a complete migration using the Yaml `UpgradeSpringBoot_3_0` correctly replaces the Spring Security dependencies and a single renamed method.

**Comment by timtebeek** [comment]
Hmm; trying this now, but I'm not seeing the same issue at first glance; I'll explore further and see where that takes me. Here's my test for now:
<details>
<summary>test showing requestMatchers renamed as part of Spring Boot 3 migration</summary>
```java
    @Test
    void migratesAllTypesOfMatchersAsPartOfSpringBootUpgrade() {
        rewriteRun(
          spec -> spec.recipe(Environment.builder()
            .scanRuntimeClasspath("org.openrewrite")
            .build()
            .activateRecipes("org.openrewrite.java.spring.boot3.UpgradeSpringBoot_3_0")),
          mavenProject( "project",
            //language=java
            srcMainJava(
              java(
                """
                  package com.example;
                                
                  import org.springframework.context.annotation.Bean;
                  import org.springframework.context.annotation.Configuration;
                  import org.springframework.http.HttpMethod;
                  import org.springframework.security.config.annotation.web.builders.HttpSecurity;
                  import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
                  import org.springframework.security.web.SecurityFilterChain;
                                
                  @Configuration
                  @EnableWebSecurity
                  class SecurityConfig {
                      @Bean
                      SecurityFilterChain securityFilterChain(HttpSecurity http) {
                          http
                                  .authorizeHttpRequests((authz) -> authz
                                          .regexMatchers("/api/admin/**").hasRole("ADMIN")
                                          .antMatchers(HttpMethod.GET, "/api/user/**").hasRole("USER")
                                          .mvcMatchers(HttpMethod.PATCH).denyAll()
                                          .anyRequest().authenticated()
                                  );
                          return http.build();
                      }
                  }
                  """
                ,
                """
                  package com.example;
                              
                  import org.springframework.context.annotation.Bean;
                  import org.springframework.context.annotation.Configuration;
                  import org.springframework.http.HttpMethod;
                  import org.springframework.security.config.annotation.web.builders.HttpSecurity;
                  import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
                  import org.springframework.security.web.SecurityFilterChain;
                              
                  @Configuration
                  @EnableWebSecurity
                  class SecurityConfig {
                      @Bean
                      SecurityFilterChain securityFilterChain(HttpSecurity http) {
                          http
                                  .authorizeHttpRequests((authz) -> authz
                                          .requestMatchers("/api/admin/**").hasRole("ADMIN")
                                          .requestMatchers(HttpMethod.GET, "/api/user/**").hasRole("USER")
                                          .requestMatchers(HttpMethod.PATCH).denyAll()
                                          .anyRequest().authenticated()
                                  );
                          return http.build();
                      }
                  }
                  """
              )
            ),
            //language=xml
            pomXml("""
              <?xml version="1.0" encoding="UTF-8"?>
              <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
              	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
              	<modelVersion>4.0.0</modelVersion>
              	<parent>
              		<groupId>org.springframework.boot</groupId>
              		<artifactId>spring-boot-starter-parent</artifactId>
              		<version>2.7.14</version>
              		<relativePath/> <!-- lookup parent from repository -->
              	</parent>
              	<groupId>com.example</groupId>
              	<artifactId>demo</artifactId>
              	<version>0.0.1-SNAPSHOT</version>
              	<name>demo</name>
              	<description>Demo project for Spring Boot</description>
              	<properties>
              		<java.version>17</java.version>
              	</properties>
              	<dependencies>
              		<dependency>
              			<groupId>org.springframework.boot</groupId>
              			<artifactId>spring-boot-starter-security</artifactId>
              		</dependency>
              	</dependencies>
              </project>
              ""","""
              <?xml version="1.0" encoding="UTF-8"?>
              <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
              	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
              	<modelVersion>4.0.0</modelVersion>
              	<parent>
              		<groupId>org.springframework.boot</groupId>
              		<artifactId>spring-boot-starter-parent</artifactId>
              		<version>3.0.9</version>
              		<relativePath/> <!-- lookup parent from repository -->
              	</parent>
              	<groupId>com.example</groupId>
              	<artifactId>demo</artifactId>
              	<version>0.0.1-SNAPSHOT</version>
              	<name>demo</name>
              	<description>Demo project for Spring Boot</description>
              	<properties>
              		<java.version>17</java.version>
              	</properties>
              	<dependencies>
              		<dependency>
              			<groupId>org.springframework.boot</groupId>
              			<artifactId>spring-boot-starter-security</artifactId>
              		</dependency>
              	</dependencies>
              </project>
              """)
          )
        );
    }
```
</details>

**Comment by timtebeek** [comment]
Sooo fun times; turns out the recipe implementation as it stands right now looks for matching replacement methods in the class loaded, but that class at the time of recipe execution is 5.7.x, whereas we need the class available to recipe execution to be 5.8.x to actually find the replacement method:
https://javadoc.io/doc/org.springframework.security/spring-security-config/5.7.10/org/springframework/security/config/annotation/web/AbstractRequestMatcherRegistry.html
https://javadoc.io/doc/org.springframework.security/spring-security-config/5.8.5/org/springframework/security/config/annotation/web/AbstractRequestMatcherRegistry.html
Will need to revise the recipe to not have that limitation.

**Comment by timtebeek** [comment]
Thanks for the detailed report @JeremyTwiggs ; this one was a bit confusing as we had tests that covered the recipe, but unfortunately at runtime the recipe failed due to unavailability of the correct version of the target class. That's now fixed with a slightly different implementation, and verified to work on a sample application at runtime.

I'll follow up with a patch release and hope that resolves the issue for you as well (if you hadn't upgraded already by now). Thanks for your patience as we went through this one. :)
